File: apps/agent-orchestrator/src/knowledge-base/probe.ts
import type { CorpusChunk, CorpusSearchResult } from "./types.js";

/**
 * The unit a provider authorizes at (docs/adr/0040).
 *
 * It differs per provider and is not a wrinkle to work around. Confluence and
 * Drive authorize a page or a file; Slack authorizes a CHANNEL, because
 * membership is the access unit and there is no per-message permission to
 * check. A per-connection provider therefore settles every candidate from that
 * connection with one probe, which is cheaper than the per-resource case rather
 * than harder.
 *
 * PARITY: `engines/temporal/internal/corpus/probe.go`.
 */
export type Granularity = "resource" | "connection";

/** Identifies what to authorize. `sourceId` is absent for connection granularity. */
export interface ProbeRequest {
  connectionId: string;
  sourceId?: string;
}

/**
 * The source's answer, and the ONLY acceptable origin for anything displayable.
 *
 * Title and URL come back here rather than from the mirror because citations
 * are content (docs/adr/0040): an answer that cites a page title the user may
 * not read has disclosed the thing while appearing to have returned nothing.
 */
export interface ProbeResult {
  allowed: boolean;
  title: string;
  url: string;
  /**
   * The version the source holds right now. A mismatch against the indexed
   * chunk means the passage is stale — still usable, while saying so, or
   * replaceable by fetching the live document.
   */
  version?: string;
}

/**
 * Asks the source, as the calling user, whether a resource is readable.
 *
 * Implementations MUST distinguish error classes: a permission failure is a
 * drop, anything else is not. Returning a transient failure as a denial makes
 * the same question return different evidence depending on whether the source
 * was busy.
 */
export interface Prober {
  granularity(connectionId: string): Granularity;
  probe(request: ProbeRequest): Promise<ProbeResult>;
}

/** The source says this user may not read it. Drop. */
export class PermissionDeniedError extends Error {
  readonly name = "PermissionDeniedError";
}

/**
 * We could not find out — a 429, a 5xx, a timeout.
 *
 * Deliberately NOT a denial. Treating it as one silently shrinks an answer in a
 * way nobody can see, and makes results non-deterministic across retries.
 */
export class TransientProbeError extends Error {
  readonly name = "TransientProbeError";
}

/**
 * A candidate the source confirmed this user may read. `title` and `url`
 * shadow anything the chunk carried: they are the probe's, not the mirror's.
 */
export interface AuthorizedChunk {
  chunk: CorpusChunk;
  score: number;
  title: string;
  url: string;
  version?: string;
  /**
   * The source has moved past the indexed version. Still readable by this user,
   * so not dropped — but an answer built on it should say how old it is, or
   * fetch the live document instead.
   */
  stale: boolean;
}

export interface AuthorizeOutcome {
  chunks: AuthorizedChunk[];
  /**
   * How many candidates the source refused. Expected and harmless: it measures
   * the mirror's staleness in the permissive direction.
   */
  denied: number;
  /**
   * Sources whose probe failed transiently. NOT dropped silently — the caller
   * surfaces them as a partial-results warning, because an answer quietly
   * missing evidence is worse than one that admits it could not check.
   */
  undetermined: string[];
}

/**
 * The gate between candidate retrieval and anything the model sees
 * (docs/adr/0040).
 *
 * The mirror got us a short list; this is where the source decides. Every
 * surviving chunk was confirmed readable, at query time, by the source system
 * itself under this user's own credentials.
 *
 * Probes are de-duplicated before they are issued — by connection for a
 * connection-granularity provider, by source otherwise — so eight chunks from
 * one page cost one probe, not eight. That de-duplication is most of why this
 * step is affordable.
 */
export async function authorize(
  prober: Prober,
  hits: CorpusSearchResult[],
): Promise<AuthorizeOutcome> {
  if (hits.length === 0) return { chunks: [], denied: 0, undetermined: [] };

  const keyFor = (chunk: CorpusChunk): string =>
    prober.granularity(chunk.connectionId) === "connection"
      ? `${chunk.connectionId}\u0000`
      : `${chunk.connectionId}\u0000${chunk.sourceId}`;

  const requests = new Map<string, ProbeRequest>();
  for (const hit of hits) {
    const key = keyFor(hit.chunk);
    if (!requests.has(key)) {
      requests.set(
        key,
        prober.granularity(hit.chunk.connectionId) === "connection"
          ? { connectionId: hit.chunk.connectionId }
          : { connectionId: hit.chunk.connectionId, sourceId: hit.chunk.sourceId },
      );
    }
  }

  const keys = [...requests.keys()];
  const settled = await Promise.allSettled(keys.map((key) => prober.probe(requests.get(key)!)));

  const results = new Map<string, PromiseSettledResult<ProbeResult>>();
  for (let i = 0; i < keys.length; i += 1) {
    results.set(keys[i]!, settled[i]!);
  }

  // An error that is neither a denial nor transient is a programming error in
  // the driver, not an authorization answer. Failing the whole search is
  // correct: guessing which it meant is how a leak gets introduced.
  for (const outcome of settled) {
    if (outcome.status === "rejected") {
      const reason: unknown = outcome.reason;
      const classified =
        reason instanceof PermissionDeniedError || reason instanceof TransientProbeError;
      if (!classified) throw reason;
    }
  }

  const chunks: AuthorizedChunk[] = [];
  const undetermined = new Set<string>();
  let denied = 0;

  for (const hit of hits) {
    const outcome = results.get(keyFor(hit.chunk))!;

    if (outcome.status === "rejected") {
      if (outcome.reason instanceof TransientProbeError) {
        undetermined.add(`${hit.chunk.connectionId}/${hit.chunk.sourceId}`);
      } else {
        denied += 1;
      }
      continue;
    }
    if (!outcome.value.allowed) {
      denied += 1;
      continue;
    }

    chunks.push({
      chunk: hit.chunk,
      score: hit.score,
      title: outcome.value.title,
      url: outcome.value.url,
      version: outcome.value.version,
      // A provider reporting no version cannot show a chunk to be stale, so do
      // not claim that it is.
      stale: Boolean(outcome.value.version) && outcome.value.version !== hit.chunk.version,
    });
  }

  return { chunks, denied, undetermined: [...undetermined].sort() };
}

File: engines/temporal/internal/corpus/probe.go
package corpus

import (
	"context"
	"errors"
	"sort"
	"sync"
)

// Granularity is the unit a provider authorizes at (ADR 0040).
//
// It differs per provider and is not a wrinkle to work around. Confluence and
// Drive authorize a page or a file; Slack authorizes a CHANNEL, because
// membership is the access unit and there is no per-message permission to
// check. A per-connection provider therefore settles every candidate from that
// connection with one probe, which is cheaper than the per-resource case rather
// than harder.
type Granularity string

const (
	GranularityResource   Granularity = "resource"
	GranularityConnection Granularity = "connection"
)

// ProbeRequest identifies what to authorize. SourceID is empty for a
// connection-granularity provider.
type ProbeRequest struct {
	CorpusID string
	SourceID     string
}

// ProbeResult is the source's answer, and the ONLY acceptable origin for
// anything displayable.
//
// Title and URL come back here rather than from the mirror because citations
// are content (ADR 0040): an answer that cites a page title the user may not
// read has disclosed the thing while appearing to have returned nothing.
type ProbeResult struct {
	Allowed bool
	Title   string
	URL     string
	// Version the source holds right now. A mismatch against the indexed chunk
	// means the passage is stale, which the answer may still use — while saying
	// so — or replace by fetching the live document.
	Version string
}

// Prober asks the source, as the calling user, whether a resource is readable.
//
// Implementations MUST distinguish error classes (ADR 0040): a permission
// failure is a drop, and anything else is not. Returning a transient failure as
// a denial makes the same question return different evidence depending on
// whether the source was busy.
type Prober interface {
	Granularity(connectionID string) Granularity
	Probe(ctx context.Context, req ProbeRequest) (ProbeResult, error)
}

// PermissionDenied means the source says this user may not read it. Drop.
type PermissionDenied struct{ Err error }

func (e *PermissionDenied) Error() string { return "permission denied: " + e.Err.Error() }
func (e *PermissionDenied) Unwrap() error { return e.Err }

// Transient means we could not find out — a 429, a 5xx, a timeout.
//
// Deliberately NOT a denial. Treating it as one silently shrinks an answer in a
// way nobody can see, and makes results non-deterministic across retries.
type Transient struct{ Err error }

func (e *Transient) Error() string { return "transient probe failure: " + e.Err.Error() }
func (e *Transient) Unwrap() error { return e.Err }

// AuthorizedChunk is a candidate the source confirmed this user may read.
//
// Title and URL shadow anything the chunk carried: they are the probe's, not
// the mirror's.
type AuthorizedChunk struct {
	Chunk   Chunk
	Score   float32
	Title   string
	URL     string
	Version string
	// Stale means the source has moved past the indexed version. The passage is
	// still readable by this user, so it is not dropped — but an answer built on
	// it should say how old it is, or fetch the live document instead.
	Stale bool
}

// AuthorizeOutcome is what survived, plus what did not and why.
type AuthorizeOutcome struct {
	Chunks []AuthorizedChunk
	// Denied is how many candidates the source refused. Expected and harmless:
	// it measures the mirror's staleness in the permissive direction.
	Denied int
	// Undetermined names sources whose probe failed transiently. These are NOT
	// dropped silently — the caller surfaces them as a partial-results warning,
	// because an answer quietly missing evidence is worse than one that admits
	// it could not check.
	Undetermined []string
}

// Authorize is the gate between candidate retrieval and anything the model
// sees (ADR 0040).
//
// The mirror got us a short list; this is where the source decides. Every
// surviving chunk was confirmed readable, at query time, by the source system
// itself under this user's own credentials.
//
// Probes are deduplicated before they are issued — by connection for a
// connection-granularity provider, by source otherwise — so eight chunks from
// one page cost one probe, not eight. That deduplication is most of why this
// step is affordable.
func Authorize(ctx context.Context, prober Prober, hits []Hit) (AuthorizeOutcome, error) {
	if len(hits) == 0 {
		return AuthorizeOutcome{}, nil
	}

	requests := map[ProbeRequest]struct{}{}
	keyFor := func(chunk Chunk) ProbeRequest {
		if prober.Granularity(chunk.CorpusID) == GranularityConnection {
			return ProbeRequest{CorpusID: chunk.CorpusID}
		}
		return ProbeRequest{CorpusID: chunk.CorpusID, SourceID: chunk.SourceID}
	}
	for _, hit := range hits {
		requests[keyFor(hit.Chunk)] = struct{}{}
	}

	type probeOutcome struct {
		result ProbeResult
		err    error
	}
	var (
		mu       sync.Mutex
		results  = make(map[ProbeRequest]probeOutcome, len(requests))
		wg       sync.WaitGroup
		fatalErr error
	)

	for req := range requests {
		wg.Add(1)
		go func(req ProbeRequest) {
			defer wg.Done()
			result, err := prober.Probe(ctx, req)

			mu.Lock()
			defer mu.Unlock()
			results[req] = probeOutcome{result: result, err: err}

			// An error that is neither a denial nor transient is a programming
			// error in the driver, not an authorization answer. Failing the
			// whole search is correct: guessing which it meant is how a leak
			// gets introduced.
			var denied *PermissionDenied
			var transient *Transient
			if err != nil && !errors.As(err, &denied) && !errors.As(err, &transient) && fatalErr == nil {
				fatalErr = err
			}
		}(req)
	}
	wg.Wait()

	if fatalErr != nil {
		return AuthorizeOutcome{}, fatalErr
	}

	outcome := AuthorizeOutcome{Chunks: make([]AuthorizedChunk, 0, len(hits))}
	undetermined := map[string]struct{}{}

	for _, hit := range hits {
		probe := results[keyFor(hit.Chunk)]

		var transient *Transient
		if errors.As(probe.err, &transient) {
			undetermined[hit.Chunk.CorpusID+"/"+hit.Chunk.SourceID] = struct{}{}
			continue
		}
		var denied *PermissionDenied
		if errors.As(probe.err, &denied) || !probe.result.Allowed {
			outcome.Denied++
			continue
		}

		outcome.Chunks = append(outcome.Chunks, AuthorizedChunk{
			Chunk:   hit.Chunk,
			Score:   hit.Score,
			Title:   probe.result.Title,
			URL:     probe.result.URL,
			Version: probe.result.Version,
			// An empty probe version means the provider does not report one; a
			// chunk cannot then be shown to be stale, so it is not claimed to be.
			Stale: probe.result.Version != "" && probe.result.Version != hit.Chunk.Version,
		})
	}

	for source := range undetermined {
		outcome.Undetermined = append(outcome.Undetermined, source)
	}
	sort.Strings(outcome.Undetermined)
	return outcome, nil
}
